BrainPortal/lib/boutiques_dir_maker.rb

```diff
@@ -0,0 +1,166 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# Some tools take expect that a file/directory input is
+# placed input a parent directory.
+#
+# For example we have the following Boutiques command-line:
+#
+#     apptool [PRECOMPUTED_INPUT]
+#
+# with the following input in Boutiques:
+#
+#     {
+#       "description": "A directory that contain the precomputed data process by...",
+#       "id": "precomputed_input",
+#       "name": "Precomputed input",
+#       "optional": false,
+#       "type": "File",
+#       "value-key": "[PRECOMPUTED_INPUT]"
+#     }
+#
+# and in the `cbrain:integrator_modules` section:
+#
+#     "BoutiquesDirMaker": {
+#         "precomputed_input": "fake_parent_dir"
+#      },
+#
+# In CBRAIN the user will select a userfile for example `sub-n`,
+# the final command line will become:
+#
+#     apptool precomputed/sub-n
+#
+module BoutiquesDirMaker
+
+  # Note: to access the revision info of the module,
+  # you need to access the constant directly, the
+  # object method revision_info() won't work.
+  Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
+
+  ############################################
+  # Portal Side Modifications
+  ############################################
+
+  # Adjust the description for the input with the mention of
+  # the fake parent directory
+  def descriptor_for_form #:nodoc:
+    descriptor = super.dup
+    boutiques_dir_maker_list = descriptor.custom_module_info('BoutiquesDirMaker')
+
+    boutiques_dir_maker_list.each do |inputid,dirname|
+      # Adjust the description
+      input        = descriptor.input_by_id(inputid)
+      description  = input.description.presence || ""
+      description += "\n" if description
+      description += "This input will be copied in a parent folder `#{dirname}`.\n The parent folder will be used in the command line"
+      input.description = description
+    end
+
+    descriptor
+  end
+
+  ############################################
+  # Bourreau (Cluster) Side Modifications
+  ############################################
+
+  # For input in the `BoutiquesDirMaker` section,
+  # create a directory based on the value BoutiquesDirMaker[input_id]
+  # that will contains a symlink to the orginal selected Userfile.
+  def setup #:nodoc:
+    return false unless super
+
+    descriptor = self.descriptor_for_setup
+    self.addlog(descriptor.file_revision_info.format("%f rev. %s %a %d"))
+
+    boutiques_dir_maker_list = descriptor.custom_module_info('BoutiquesDirMaker')
```

Comment:
Why is this variable called `_list` when it's a hash?

Reply:
When naming a hash, it is customary to name it after a description of the VALUES in it. E.g.

```ruby
fruits = { banana: 'yellow', apple: 'red' } # BAD name
colors = { banana: 'yellow', apple: 'red' } # BETTER name

# When accessing, it makes reading the code easier when the hash is named after the values:
color = fruits['banana'] # BAD
color = colors['banana'] # BETTER
```